gflownet/utils/batch.py

```diff
@@ -0,0 +1,175 @@
+                )
+                self.parents_actions = torch.cat(
+                    [
+                        tfloat(x, device=self.device, float=self.float)
+                        for x in self.parents_actions
+                    ]
+                )
+            elif self.loss == "trajectorybalance":
+                self.mask_invalid_actions_backward = tbool(
+                    self.mask_invalid_actions_backward, device=self.device
+                )
+            self._process_parents()
+
+    def _process_states(self):
+        self.state_gfn = tfloat(self.state, device=self.device, float=self.float)
+        states = []
+        for state, env_id in zip(self.state, self.env_id):
+            states.append(self.envs[env_id].state2policy(state))
+        self.state = tfloat(states, device=self.device, float=self.float)
+
+    def _process_parents(self):
+        if self.loss == "flowmatch":
+            parents = []
+            for par, env_id in zip(self.parents, self.env_id):
+                parents.append(
+                    tfloat(
+                        self.envs[env_id.item()].statebatch2policy(par),
+                        device=self.device,
+                        float=self.float,
+                    )
+                )
+            self.parents = torch.cat(parents)
+        elif self.loss == "trajectorybalance":
+            parents = torch.zeros_like(self.state)
+            for env_id, traj in self.trajectory_indicies.items():
+                parents[traj[0]] = tfloat(
+                    self.envs[env_id].state2policy(self.envs[env_id].source),
+                    device=self.device,
+                    float=self.float,
+                )
+                parents[traj[1:]] = self.state[traj[:-1]]
+            self.parents = parents
+
+    def merge(self, another_batch):
+        if self.loss != another_batch.loss:
+            raise Exception("Cannot merge batches with different losses")
+        self.envs.update(another_batch.envs)
+        self.state += another_batch.state
+        self.action += another_batch.action
+        self.done += another_batch.done
+        self.env_id += another_batch.env_id
+        self.mask_invalid_actions_forward += another_batch.mask_invalid_actions_forward
+        self.mask_invalid_actions_backward += (
+            another_batch.mask_invalid_actions_backward
+        )
+        self.parents += another_batch.parents
+        self.parents_actions += another_batch.parents_actions
+        self.step += another_batch.step
+
+    def _process_trajectory_indicies(self):
+        trajs = {env_id: [] for env_id in self.envs.keys()}
+        for idx, (env_id, step) in enumerate(zip(self.env_id, self.step)):
+            trajs[env_id.item()].append((idx, step))
+        trajs = {
+            env_id: list(map(lambda x: x[0], sorted(traj, key=lambda x: x[1])))
+            for env_id, traj in trajs.items()
+        }
+        self.trajectory_indicies = trajs
```

Comment:
Sasha, would mind writing the docstring with what this method is expected to do? It's got 3 for loops and it is a bit hard to understand. I am wondering if we could simplify this.

Reply:
Sure, I can do that. It basically creates a dict where keys are env_ids (which are also trajectory ids as we assume that one env corresponds to only one trajectory in the batch) and values are indexes of the states in self.states which correspond to this trajectory (and they go in the right order from s_1 to s_f). 
As for the loops, the fist one is to create an empty dict, we can probably eliminate it with the defaultdict. The second loop gathers all the indexes for each trajectory together with step numbers (for sorting them later). And the last one sorts the indexes according to steps and removes steps.